### DISCUSSION TITLE
Unable to Stop Event Propagation on DataGridRow with PointerPressedEvent

### DISCUSSION THREAD

**Discussion by buchmiet** [AvaloniaUI/Avalonia#13090] [General]
I'm working on an Avalonia project where I have a DataGrid with CheckBox controls within its rows. I want to suppress the bubbling of the PointerPressedEvent on the CheckBox to prevent other UI elements (specifically the DataGridRow) from reacting to the click event.

In WPF, I would typically handle this scenario with a custom Behavior class that hooks into the PreviewMouseLeftButtonDown event, but in Avalonia, the analogous event would be PointerPressedEvent. Here's an outline of the custom behavior I've attempted to implement in Avalonia:

```
public class SuppressMouseBehavior : Behavior<CheckBox>
{
    protected override void OnAttached()
    {
        base.OnAttached();
        AssociatedObject.AddHandler(InputElement.PointerPressedEvent, CheckBox_PointerPressed, handledEventsToo: true);
    }

    private void CheckBox_PointerPressed(object sender, PointerPressedEventArgs e)
    {
        e.Handled = true; // in WPF that stops event propagating, so clicking on a checkbox in a 
                                    // checkbox does not fire RowDetailsTemplate from the underlying DataGridRow
                                    // here e.Handled is already true anyway
        // Code to toggle the CheckBox state
    }
}
```

I've attached this behavior to the CheckBox controls within my DataGrid. However, when I click on a CheckBox, the PointerPressedEvent still propagates to the DataGridRow, causing undesired behaviors such as expanding or collapsing the row details.

I've also attempted to hook into the PointerPressedEvent at the DataGridRow level and set the e.Handled property to true, but this didn't stop the event propagation either. Below is the snippet for handling PointerPressedEvent on DataGridRow:

```
private void OnRowPointerPressed(object sender, PointerPressedEventArgs e)
{
    e.Handled = true;  // This line is reached in the debugger, but event propagation isn't stopped
}
```

I've used AssociatedObject.AddHandler with handledEventsToo: true to ensure my event handler gets called, but setting e.Handled to true doesn't seem to stop the event propagation as expected.

I need to find a way to stop the PointerPressedEvent propagation on the CheckBox to ensure that clicking the CheckBox doesn't trigger any actions on the underlying DataGridRow. Any insights or suggestions on how to achieve this in Avalonia would be greatly appreciated.

**Comment by maxkatz6** [comment]
> In WPF, I would typically handle this scenario with a custom Behavior class that hooks into the PreviewMouseLeftButtonDown event, but in Avalonia, the analogous event would be PointerPressedEvent

Not quite, in Avalonia the analogous event is PointerPressedEvent with Tunnel handling strategy. You can specify the handling strategy as an argument in AddHandler method.

**Reply by buchmiet** [reply]
Thanks for the previous tip. I gave it a shot and tried using the Tunnel handling strategy for the PointerPressedEvent in Avalonia with the AddHandler method. Yet, I'm still running into the same wall where the RowDetailsTemplate pops up whenever I click a CheckBox in a DataGridRow. Here's the code following your suggestion:

public class SuppressMouseBehavior : Behavior<CheckBox>
{
    protected override void OnAttached()
    {
        base.OnAttached();
        AssociatedObject.AddHandler(
            InputElement.PointerPressedEvent, 
            CheckBox_PointerPressed, 
            RoutingStrategies.Tunnel, 
            handledEventsToo: true
        );
    }

    private void CheckBox_PointerPressed(object sender, PointerPressedEventArgs e)
    {
        e.Handled = true; // Trying to halt event propagation here
    }
}

This changes the Despite my effort to halt the event propagation by setting e.Handled = true in the CheckBox_PointerPressed method, the PointerPressedEvent seems to keep on trucking, and the RowDetailsTemplate still shows up.

Would you have any more insights or tricks up your sleeve on how to keep the RowDetailsTemplate from showing when a CheckBox within a DataGridRow is clicked ?

much appreciated!

**Reply by maxkatz6** [reply]
You can use DevTools to see which event was handled and where (on which control). Should be helpful.
https://docs.avaloniaui.net/docs/getting-started/developer-tools

**Reply by buchmiet** [reply]
Thanks Max!

I wasn't able to harness the events as wherever I was depressing them, some other control all the way up to the MainWindow was propagating them down to DataGrid. I have found another solution - in case anyone will be trying to figure out a way how to exclude a checkbox or other Control from triggering displaying RowDetails, here is the behaviour:

  ```
public class RowEventBehavior : Behavior<DataGrid>
  {
      protected override void OnAttached()
      {
          base.OnAttached();         
          AssociatedObject.LoadingRow += OnLoadingRow;         
      }

      protected override void OnDetaching()
      {
          AssociatedObject.LoadingRow -= OnLoadingRow;
          base.OnDetaching();
      }

      private void OnLoadingRow(object sender,  DataGridRowEventArgs e)
      {
          e.Row.AddHandler(DataGridRow.PointerPressedEvent, OnRowPointerPressed, RoutingStrategies.Tunnel, handledEventsToo: true);
      }
     

      private void OnRowPointerPressed(object sender, PointerPressedEventArgs e)
      {            
          var PressedSource = e.Source as Control;
          if (PressedSource.TemplatedParent is CheckBox)
          {
              AssociatedObject.RowDetailsVisibilityMode= DataGridRowDetailsVisibilityMode.Collapsed;
          } else
          {
              AssociatedObject.RowDetailsVisibilityMode = DataGridRowDetailsVisibilityMode.VisibleWhenSelected;
          }           
      }
```
